language_bindings/rust/rust_build/src/lib.rs

```diff
@@ -0,0 +1,360 @@
+        cmd.args([
+            "-r",
+            "runtime",
+            "-o",
+            self.rs_output.to_str().unwrap(),
+            "target=host",
+        ]);
+        cmd.env("LD_LIBRARY_PATH", self.halide_path.join("lib"));
+        let output = cmd.output().expect("failed to make runtime");
+        let result;
+        if output.status.success() {
+            result = self.rename_runtime();
+            if result.is_ok() {
+                println!("cargo:rustc-link-lib=static={}", "runtime");
+            }
+        } else {
+            let errors =
+                String::from("failed to make runtime with gen: ").add(self.gen_name.as_str());
+            result = Err(io::Error::new(ErrorKind::Other, errors.as_str()));
+        }
+        result
+    }
+
+    fn rename_runtime(&self) -> Result<()> {
+        println!(
+            "file name: {:?}",
+            self.rs_output
+                .join("runtime")
+                .with_extension("a")
+                .to_str()
+                .unwrap()
+        );
+        println!(
+            "file out: {:?}",
+            self.rs_output
+                .join(String::from("lib").add("runtime"))
+                .with_extension("a")
+                .to_str()
+                .unwrap()
+        );
+        fs::rename(
+            self.rs_output
+                .join("runtime")
+                .with_extension("a")
+                .to_str()
+                .unwrap(),
+            self.rs_output
+                .join(String::from("lib").add("runtime"))
+                .with_extension("a")
+                .to_str()
+                .unwrap(),
+        )
+    }
+
+    /// Renames the gen outputs to be what rust and bindgen expect on linux distros
+    ///
+    /// IE halide.a -> libhalide.a
+    pub fn rename(&self) -> Result<()> {
+        println!(
+            "file name: {:?}",
+            self.rs_output
+                .join(self.gen_name.as_str())
+                .with_extension("a")
+                .to_str()
+                .unwrap()
+        );
+        println!(
+            "file out: {:?}",
+            self.rs_output
+                .join(String::from("lib").add(self.gen_name.as_str()))
+                .with_extension("a")
+                .to_str()
+                .unwrap()
+        );
+        fs::rename(
+            self.rs_output
+                .join(self.gen_name.as_str())
+                .with_extension("a")
+                .to_str()
+                .unwrap(),
+            self.rs_output
+                .join(String::from("lib").add(self.gen_name.as_str()))
+                .with_extension("a")
+                .to_str()
+                .unwrap(),
+        )
+    }
+
+    /// Run bindgen on a generator
+    ///
+    /// This bind only creates the genname funtion binding and specificly blocklists the buffer_t
+    pub fn bind(&self) -> Result<()> {
+        let bindings = bindgen::Builder::default()
```

Comment:
In the example, this seems to be used only to produce a single, simple `iir_blur` function declaration. `bindgen` is a very heavy dependency as it requires LLVM, and rather fragile as it's sensitive to the C/C++ build environment. If the minimal requirements in the example are representative, could `bindgen` be avoided by just having downstream code declare the interface in question themselves? This would also make downstream code more readable, as right now it's unclear what the signature of `iir_blur` is, exactly.

Reply:
We were under the impression that LLVM would already be a requirement for Halide. bindgen is used here to create an iir_blur.rs file with a single iir_blur(buffer_t, float, buffer_t) extern function. This matches the function in iir_blur.h while we could write a function that takes a name and generates an extern function .rs in the same format it would not be able to deal with other function signatures very well. If for example, the generator required an int, 2 floats, or an additional buffer.

Reply:
> We were under the impression that LLVM would already be a requirement for Halide

Even if LLVM is a dependency of Halide, that does not necessarily imply that the user has a compatible version of libclang installed somewhere that bindgen can find it. LLVM versions aren't all compatible, LLVM doesn't necessarily imply libclang, and It may even be possible to have Halide available with no independent LLVM on the system at all, if Halide statically links it.

> while we could write a function that takes a name and generates an extern function .rs in the same format it would not be able to deal with other function signatures very well. If for example, the generator required an int, 2 floats, or an additional buffer.

Because this is a single function with a very predictable signature, it may be reasonable to ask users to write the FFI declaration themselves. It would take about as much space as the `include!` invocation, and make the process overall much simpler and lighter. You could even provide a declarative macro that generates both the FFI declaration and a safe wrapper.